torch/distributed/fsdp/_optim_utils.py

```diff
@@ -1071,51 +1070,104 @@ def _get_param_to_param_id_from_optim_input(
     return {param: param_id for param_id, param in enumerate(param_id_to_param)}
 
 
-def _get_unflat_to_flat_param_ids(
-    flat_to_unflat_param_ids: Dict[int, List[int]],
-) -> List[int]:
-    """
-    Inverts the mapping ``flat_to_unflat_param_ids`` to be from unflattened
-    parameter ID to flattened parameter ID, where the unflattened parameter ID
-    is the index in the returned :class:`list`. There may be multiple
-    unflattened parameter IDs mapping to the same flattened parameter ID.
+def _is_zero_dim_tensor(x: Any) -> bool:
+    return torch.is_tensor(x) and x.dim() == 0
 
-    Args:
-        flat_to_unflat_param_ids (Dict[int, List[int]]): A mapping from
-            flattened parameter ID to a :class:`list` of corresponding
-            unflattened parameter IDs.
 
-    Returns:
-        List[int]: A mapping from unflattened parameter ID to flattened
-        parameter ID, where the unflattened parameter ID is the index in the
-        :class:`list`.
+def _map_param_id_to_optim_keys(
+    optim_state_dict: Dict[str, Any],
+    group: Optional[dist.ProcessGroup],
```

Comment:
General comment: While refactoring optimizer state dict, it could be worthwhile to keep an eye out for process group usage (mainly looking for any accidental default process group usage). This is in preparation for HSDP, where every process group usage now should map to the intra-node process group from HSDP.

cc: @rohan-varma